static/app/views/dashboards/manage/index.tsx

```diff
@@ -63,18 +66,74 @@ type State = {
   dashboards: DashboardListItem[] | null;
   dashboardsPageLinks: string;
   showTemplates: boolean;
+  windowWidth: number;
 } & DeprecatedAsyncView['state'];
 
 class ManageDashboards extends DeprecatedAsyncView<Props, State> {
+  shouldReload = true;
+
+  constructor(props: Props, context: any) {
+    super(props, context);
+
+    this.state = {
+      ...super.getDefaultState(),
+      ...this.getDefaultState(),
+      dashboards: [],
+      dashboardsPageLinks: '',
+      windowWidth: window.innerWidth,
+    };
+  }
+
   getDefaultState() {
     return {
       ...super.getDefaultState(),
       showTemplates: shouldShowTemplates(),
     };
   }
 
+  debouncedHandleResize = debounce(() => {
+    const currentWidth = window.innerWidth;
+    // Only update state if the width has changed
+    if (currentWidth !== this.state.windowWidth) {
+      this.setState({windowWidth: currentWidth});
+    }
+    const paginationObject = parseLinkHeader(this.state.dashboardsPageLinks);
+    if (
+      this.state.dashboards?.length &&
+      paginationObject.next.results &&
+      this.getDashboardsPerPage() > this.state.dashboards.length
+    ) {
+      this.reloadData();
+    }
+  }, 250);
+
+  componentDidMount() {
+    window.addEventListener('resize', this.debouncedHandleResize);
+    super.componentDidMount();
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.debouncedHandleResize);
+    super.componentWillUnmount();
+  }
+
+  shouldComponentUpdate(nextProps: Readonly<Props>, nextState: Readonly<State>): boolean {
+    const propsEqual = isEqual(nextProps, this.props);
+    const stateEqual = isEqual(nextState, this.state);
+    const windowEqual = isEqual(this.state.windowWidth, window.innerWidth);
+    return !propsEqual || !stateEqual || !windowEqual;
+  }
+
+  getDashboardsPerPage(): number {
+    // min-midth of widget is 300px with additional 16px margin on each side (also accounting for side menu)
+    return Math.floor((window.innerWidth - 100) / (300 + 32)) < 3
+      ? 8
+      : Math.floor((window.innerWidth - 100) / (300 + 32)) * 3;
+  }
+
```

Comment:
I think having this calculation logic separate from the layout logic is a bit of a risk. Here the code decides how many dashboards will fit based on the screen, but the dashboard elements are laid out using CSS grid that has its own logic.

If someone changes the grid CSS, this calculation might not match it anymore, and vice versa. Also, right now at some screen widths the grid is "ragged", it has empty spaces in the bottom right.

I think a more robust approach would be something like:

1. Extract the padding (32px), widget width (300), and minimum count (8) as constants in JavaScript
2. In the JavaScript here, calculate how many columns and rows of dashboards you want to render
3. Save the current row and column count on the state
4. Pass `rowCount` and `columnCount` to `DashboardList`
5. In `DashboardList` set the CSS grid column count according to `columnCount` prop
6. In `DashboardList`, add placeholder elements based on whether `columnCount * rowCount` is more than `dashboards.length`

This set up is clearer, I think! It will make sure the grid is always "perfect" (every row is full) as much as possible, and that the layout and the dashboard count are always in sync